### PR TITLE
Fix: string to integer cast

### LINKED ISSUE #5328: String to HUGEINT cast bug
### What happens?

Casting from string to hugeint behaves incorrect.

### To Reproduce

```sql
D  select '1.8259857912588366e+37'::hugeint;
┌───────────────────────────────────────────┐
│ CAST('1.8259857912588366e+37' AS HUGEINT) │
│                  int128                   │
├───────────────────────────────────────────┤
│    20000000000000000000000000000000000000 │
└───────────────────────────────────────────┘
```
But this works.
```sql
D select 1.8259857912588366e+37::hugeint;
┌─────────────────────────────────────────┐
│ CAST(1.8259857912588366e+37 AS HUGEINT) │
│                 int128                  │
├─────────────────────────────────────────┤
│  18259857912588365870837119913054699520 │
└─────────────────────────────────────────┘
```

### OS:

iOS

### DuckDB Version:

master

### DuckDB Client:

CLI

### Full Name:

Tania Bogatsch

### Affiliation:

DuckDB Labs

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by nickgerrets**
fixes #5328 

This PR adds better support for casting from string to integers (including `HUGEINT`) when exponents and/or decimals are involved.

Examples:
`'1.234e3'::INTEGER` was previously evaluated to `1000` and is now evaluated to `1234`.
`'125e-1'::INTEGER` was previously evaluated to `12` and is now evaluated to `13`.
`'150e-1'::TINYINT` was previously considered a Conversion Error and is now evaluated to `15`.

### CODE REVIEW

File: src/common/operator/cast_operators.cpp [hunk 1]
```diff
@@ -835,49 +852,176 @@ struct IntegerCastOperation {
...
-		return true;
+		using result_t = typename T::ResultType;
+		using store_t = typename T::StoreType;
+
+		result_t tmp;
+		if (!TryCast::Operation<store_t, result_t>(state.result, tmp)) {
+			return false;
+		}
+
+		while (state.decimal > 10) {
```
Comment (Tishj) [inline src/common/operator/cast_operators.cpp:1006]: We can do a single modulo 10 instead right?

**Comment by Tishj** [comment]
I tried
```sql
select '0.000005'::INTEGER
```
Which gave me `1`
I'm not sure that's correct?

**Comment by Tishj** [comment]
As a follow up, this isn't an easy question to answer, but since you're working on this I figured I'd ask you.

Out of these rounding modes, which one does DuckDB follow?
```
rounding ['TIE_TO_EVEN', 'TIE_AWAY_FROM_ZERO', 'TRUNCATE', 'CEILING', 'FLOOR']
```
I think `TIE_AWAY_FROM_ZERO` right?
That seems to be the current behavior, and what I've seen from reviewing this code, decimal `>= 5` is how we determine if we should round up.

**Comment by nickgerrets** [comment]
> Out of these rounding modes, which one does DuckDB follow?

@Tishj  `TIE_AWAY_FROM_ZERO` is indeed the intended behaviour.

File: test/sql/cast/string_to_integer_exponent_cast.test [hunk 2]
```diff
@@ -0,0 +1,143 @@
+# name: test/sql/cast/string_to_integer_exponent_cast.test
+# description: String to Integer casts with Decimals and Exponents
+# group: [cast]
```
Comment (Tishj) [inline test/sql/cast/string_to_integer_exponent_cast.test:3]: I thought of another test where we create the minimal/max value and prefix it with `0.000...`, then construct the exponent to restore the value from the fraction

But that seemed to trigger an assertion
```sql
D select '0.170141183460469231731687303715884105727e39'::HUGEINT;
┌─────────────────────────────────────────────────────────────────┐
│ CAST('0.170141183460469231731687303715884105727e39' AS HUGEINT) │
│                             int128                              │
├─────────────────────────────────────────────────────────────────┤
│                         170141183460469231731687303715884105727 │
└─────────────────────────────────────────────────────────────────┘
D select '0.0170141183460469231731687303715884105727e40'::HUGEINT;
Error: INTERNAL Error: Assertion triggered in file "/Users/thijs/DuckDBLabs/duckdb/src/common/types/hugeint.cpp" on line 314: !(rhs.upper == 0 && rhs.lower == 0)
```

**Review by Tishj — CHANGES_REQUESTED** [review]
Found a crash that should be fixed and I've requested a couple more tests

**Review by Tishj — APPROVED** [review]
Thanks I think this is ready to be merged now 👍

File: src/common/operator/cast_operators.cpp [hunk 3]
```diff
@@ -835,49 +852,180 @@ struct IntegerCastOperation {
...
+	}
+
+	template <class T, bool NEGATIVE>
+	static bool HandleExponent(T &state, int32_t exponent) {
+		using store_t = typename T::StoreType;
+
+		int32_t e = exponent;
+		// Negative Exponent
+		if (e < 0) {
+			while (e++ < 0) {
```
Comment (Mytherin) [inline src/common/operator/cast_operators.cpp:934]: Can we do an early-out when `state.result==0`

File: src/common/operator/cast_operators.cpp [hunk 4]
```diff
@@ -835,49 +852,180 @@ struct IntegerCastOperation {
...
+
+		// Positive Exponent
+		while (e-- > 0) {
+			if (!TryMultiplyOperator::Operation(state.result, (store_t)10, state.result)) {
 				return false;
 			}
-			state.result--;
+		}
+
+		if (state.decimal == 0) {
```
Comment (Mytherin) [inline src/common/operator/cast_operators.cpp:952]: Can we add tests:

```sql
select '15123456789e-2147483648'::int;
select '0e2147483647'::int;
```

File: src/common/operator/cast_operators.cpp [hunk 5]
```diff
@@ -835,49 +852,180 @@ struct IntegerCastOperation {
...
+			return Finalize<T, NEGATIVE>(state);
+		}
+
+		// Handle decimals
+		e = exponent - state.decimal_digits;
+		store_t remainder = 0;
+		if (e < 0) {
+			store_t power = 1;
+			while (e++ < 0) {
+				power *= 10;
```
Comment (Mytherin) [inline src/common/operator/cast_operators.cpp:962]: What if power goes out of bounds here, can we check this?

File: src/common/operator/cast_operators.cpp [hunk 6]
```diff
@@ -835,49 +852,180 @@ struct IntegerCastOperation {
...
-		// if digit >= 5, then we round up (or down in case of negative numbers)
-		auto increment = digit >= 5;
-		if (!increment) {
-			return true;
+		state.result = state.result * 2 + digit;
+		return true;
+	}
+
+	template <class T, bool NEGATIVE>
+	static bool HandleExponent(T &state, int32_t exponent) {
```
Comment (Mytherin) [inline src/common/operator/cast_operators.cpp:928]: Can we make exponents a `int16_t`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some more comments:

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**